### PR TITLE
Hapi Model add accumulate gradients option

### PR DESCRIPTION
**Pull request by lyuwenyu**
### PR types

New features
### PR changes

Others
### Describe

- add `accumulate` option in `Model.fit`
  - the number of steps of accumulating gradients, used to mimic large batch size
  - **only support dygraph**

- usage
```python
import paddle
import paddle.nn as nn
import paddle.vision.transforms as T

device = paddle.set_device('cpu') # or 'gpu'

net = nn.Sequential(nn.Flatten(1), nn.Linear(784, 200), nn.Tanh(), nn.Linear(200, 10))

model = paddle.Model(net)
optim = paddle.optimizer.SGD(learning_rate=1e-3, parameters=model.parameters())
model.prepare(optim, paddle.nn.CrossEntropyLoss(), paddle.metric.Accuracy())

transform = T.Compose([ T.Transpose(), T.Normalize([127.5], [127.5])])
data = paddle.vision.datasets.MNIST(mode='train', transform=transform)

model.fit(data, epochs=5, batch_size=4, verbose=1, accumulate=4)

```

### CODE REVIEW

File: python/paddle/hapi/model.py [hunk 1]
```diff
@@ -2029,8 +2042,13 @@ def _run_one_epoch(self, data_loader, callbacks, mode, logs={}):
             callbacks.on_batch_begin(mode, step, logs)
 
             if mode != 'predict':
-                outs = getattr(self, mode + '_batch')(data[:len(self._inputs)],
-                                                      data[len(self._inputs):])
+
+                _inputs = [data[:len(self._inputs)], data[len(self._inputs):]]
+                if mode == 'train':
+                    _inputs.append((step + 1) % self._accumulate == 0)
```
Comment (LiuChiachi) [inline python/paddle/hapi/model.py]: Does the last step need to be updated?

File: python/paddle/hapi/model.py [hunk 2]
```diff
@@ -1029,6 +1032,7 @@ def train_batch(self, inputs, labels=None):
                 a numpy array or paddle.Tensor, or a list of arrays or tensors 
                 (in case the model has multiple labels). If has no labels, 
                 set None. Default is None.
+            update (bool): Whether update parameters after loss.backward() computing. Using it to accumulate gradients. Default is True.
```
Comment (LiuChiachi) [inline python/paddle/hapi/model.py]: Maybe doc string should not exceed 80 columns:)

File: python/paddle/tests/test_model.py [hunk 3]
```diff
@@ -718,6 +718,21 @@ def test_dygraph_export_deploy_model_about_inputs(self):
...
+        optim = fluid.optimizer.SGD(learning_rate=0.001,
+                                    parameter_list=net.parameters())
+        inputs = [InputSpec([None, dim], 'float32', 'x')]
+        labels = [InputSpec([None, 1], 'int64', 'label')]
+        model = Model(net, inputs, labels)
+        model.prepare(optim, loss=CrossEntropyLoss(reduction="sum"))
+        loss1, = model.train_batch([data], [label], update=False)
+        loss2, = model.train_batch([data], [label], update=True)
+        np.testing.assert_almost_equal(loss1, loss2, decimal=4)
+
```
Comment (LiuChiachi) [inline python/paddle/tests/test_model.py:750]: Could you please add a test under AMP here? Thanks:)
```
model.prepare(optim, loss=CrossEntropyLoss(reduction="sum"), amp_configs='O1')

Comment (lyuwenyu) [reply python/paddle/tests/test_model.py:750]: done

File: python/paddle/hapi/model.py (hunk 2 above)
Comment (lyuwenyu) [reply python/paddle/hapi/model.py]: done

File: python/paddle/hapi/model.py (hunk 1 above)
Comment (lyuwenyu) [reply python/paddle/hapi/model.py]: Yep, I think should handle it here.  And done.

File: python/paddle/hapi/model.py [hunk 4]
```diff
@@ -1017,7 +1020,7 @@ def __init__(self, network, inputs=None, labels=None):
         else:
             self._adapter = StaticGraphAdapter(self)
 
-    def train_batch(self, inputs, labels=None):
+    def train_batch(self, inputs, labels=None, update=True):
         """
         Run one training step on a batch of data.
```
Comment (jzhang533) [inline python/paddle/hapi/model.py]: 这个docstring是否也更新一下。

File: python/paddle/hapi/model.py [hunk 5]
```diff
@@ -1537,7 +1542,8 @@ def fit(
             drop_last=False,
             shuffle=True,
             num_workers=0,
-            callbacks=None, ):
+            callbacks=None,
+            accumulate=1, ):
```
Comment (jzhang533) [inline python/paddle/hapi/model.py]: `accumulate` 是否有更加合适的名称呢？我看pytorch lignhting里叫这个：[accumulate_grad_batches](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.trainer.trainer.html#pytorch_lightning.trainer.trainer.Trainer.params.accumulate_grad_batches)

File: python/paddle/hapi/model.py [hunk 6]
```diff
@@ -1580,7 +1586,10 @@ def fit(
             callbacks (Callback|None): A list of `Callback` instances to apply
                 during training. If None, `ProgBarLogger` and `ModelCheckpoint`
                 are automatically inserted. Default: None.
-
+            accumulate (int): The number of steps to accumulate gradident during
```
Comment (jzhang533) [inline python/paddle/hapi/model.py]: steps 改成 batches是不是更好？

**Review by jzhang533 — COMMENTED** [review]
这些也麻烦顺便清理一下。
python/paddle/hapi/model.py:28:1: F401 'collections.Iterable' imported but unused
python/paddle/hapi/model.py:35:1: F401 'paddle.fluid.framework.ParamBase' imported but unused
python/paddle/hapi/model.py:36:1: F401 'paddle.fluid.framework._current_expected_place' imported but unused
python/paddle/hapi/model.py:43:1: F401 'paddle.fluid.dygraph.dygraph_to_static.program_translator.ProgramTranslator' imported but unused
python/paddle/hapi/model.py:44:1: F401 'paddle.fluid.dygraph.dygraph_to_static.program_translator.FunctionSpec' imported but unused
python/paddle/hapi/model.py:53:1: F401 'paddle.fluid.executor.scope_guard' imported but unused
python/paddle/hapi/model.py:54:1: F401 'paddle.fluid.executor.Executor' imported but unused
python/paddle/hapi/model.py:55:1: F401 'paddle.fluid.dygraph.layers.Layer' imported but unused

File: python/paddle/hapi/model.py (hunk 6 above)
Comment (lyuwenyu) [reply python/paddle/hapi/model.py]: done

File: python/paddle/hapi/model.py (hunk 4 above)
Comment (lyuwenyu) [reply python/paddle/hapi/model.py]: done

File: python/paddle/hapi/model.py (hunk 5 above)
Comment (lyuwenyu) [reply python/paddle/hapi/model.py]: done

File: python/paddle/tests/test_model.py [hunk 7]
```diff
@@ -718,6 +718,29 @@ def test_dygraph_export_deploy_model_about_inputs(self):
         model.save(save_dir, training=False)
         shutil.rmtree(save_dir)
 
+    def test_accumulate(self, ):
```
Comment (qingqing01) [inline python/paddle/tests/test_model.py:721]: better to check gradients between  one large batch size and multiple small batch sizes

**Review by jzhang533 — APPROVED** [review]
LGTM

File: python/paddle/tests/test_model.py (hunk 7 above)
Comment (lyuwenyu) [reply python/paddle/tests/test_model.py:721]: done

**Comment by jzhang533** [comment]
this warning has also been removed when `import paddle` by landing this PR
``
/usr/local/lib/python3.9/site-packages/paddle/hapi/model.py:28: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.10 it will stop working
``
